smartapps/curb/curb-energy-monitor.src/curb-energy-monitor.groovy

```diff
@@ -0,0 +1,477 @@
+    singleInstance: true
+) {
+    appSetting "clientId"
+    appSetting "clientSecret"
+}
+
+preferences {
+    page(
+        name: "auth",
+        title: "Curb",
+        nextPage: "",
+        content: "authPage",
+        uninstall: true)
+}
+
+mappings {
+    path("/oauth/initialize") {
+        action: [GET: "oauthInitUrl"]
+    }
+    path("/oauth/callback") {
+        action: [GET: "callback"]
+    }
+}
+
+def installed() {
+    log.debug "Installed with settings: ${settings}"
+}
+
+def updated() {
+    log.debug "Updated with settings: ${settings}"
+    removeChildDevices(getChildDevices())
+    unsubscribe()
+    initialize()
+}
+
+def initialize() {
+    log.debug "Initializing"
+    unschedule()
+    refreshAuthToken()
+    updateSelectedLocationId()
+    getDevices()
+    runEvery1Minute(getAllData)
+}
+
+def uninstalled() {
+    log.debug "Uninstalling"
+    removeChildDevices(getChildDevices())
+}
+
+def authPage() {
+    if (!state.accessToken) {
+        state.accessToken = createAccessToken()
+    }
+
+    if (state.authToken) {
+        return dynamicPage(name: "auth", title: "Login Successful", nextPage: "", install: true, uninstall: true) {
+            section() {
+                paragraph("Select your CURB Location")
+                input(
+                    name: "curbLocation",
+                    type: "enum",
+                    title: "CURB Location",
+                    options: state.locationNames
+                )
+            }
+        }
+    } else {
+        return dynamicPage(name: "auth", title: "Login", nextPage: "", uninstall: false) {
+            section() {
+                paragraph("Tap below to log in to the CURB service and authorize SmartThings access")
+                href url: buildRedirectUrl, style: "embedded", required: true, title: "CURB", description: "Click to enter CURB Credentials"
+            }
+        }
+    }
+}
+
+def oauthInitUrl() {
+    state.oauthInitState = UUID.randomUUID().toString()
+    def oauthParams = [
+        response_type: "code",
+        scope: "offline_access",
+        audience: "app.energycurb.com/api",
+        client_id: appSettings.clientId,
+        connection: "Users",
+        state: state.oauthInitState,
+        redirect_uri: callbackUrl
+    ]
+    redirect(location: "${curbLoginUrl}?${toQueryString(oauthParams)}")
+}
+
+def callback() {
+    def code = params.code
+    def oauthState = params.state
+    if (oauthState == state.oauthInitState) {
+        def tokenParams = [
+            grant_type: "authorization_code",
+            code: code,
+            client_id: appSettings.clientId,
+            client_secret: appSettings.clientSecret,
+            redirect_uri: callbackUrl
+        ]
+        httpPostJson([uri: curbTokenUrl, body: tokenParams]) {
+            resp ->
+                state.refreshToken = resp.data.refresh_token
+              state.authToken = resp.data.access_token
+
+            getCurbLocations()
```

Comment:
Remove this call and put it in the `authPage` method before returning the `dynamicPage`.